### PR TITLE
06elements

### PR DESCRIPTION
**Pull request by Xliff**
OK, there are quite a few changes, and I've been using the 06elements.t test as a guide. For one thing, I wonder if we really need Dom.pm. I started off porting that from Perl5's XML::LibXML. Early on I had questions with regard to the XS way of handling nodes. Maybe this is something we can leave to valgrind at a later state?

Also, I am wondering if it might not be better to merge XML::LibXML::Common into XML::LibXML::Nodish? This leaves the XML::LibXML::Node class with almost no methods, but given the way things are arranged in CStruct.pm, we'd have problems with class issues (example, XML::LibXML::Element needs some of the methods in XML::LibXML::Node, however ::Element can't be a subclass of ::Node due to the difference in structures -- xmlNode vs xmlElement)

There's also the issue of how to deal with object comparisons. It would be nice if we could just compare containers, however my implementation of isSameNode() may be lacking.

So let me know if things are moving in the direction and if I need to change things.

### CODE REVIEW

File: t/06elements-port.t [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+            my $caught = 0;
+            CATCH {
+                # TEST*$badnames
+                ok True, "setAttributeNS throws an exception for '$name'";
+                $caught = 1;
+                .resume;
+            }
+            $elem.setAttributeNS( Nil, $name, 'X' );
+            nok True, "setAttributeNS did not throw an exception for '$name'"
+                if $caught == 0;
```
Comment (FROGGS) [inline t/06elements-port.t]: better written as: unless $caught

File: lib/XML/LibXML/Attr.pm [hunk 2]
```diff
@@ -3,16 +3,22 @@ use nqp;
...
+# cw: Constantly calling refresh may kill performance... so is there a way to check
+#     if the underlying object really needs it?
+
 method name() {
+    nqp::nativecallrefresh(self);
     self.ns && self.ns.name ?? self.ns.name ~ ':' ~ self.localname !! self.localname
 }
 
 method value() {
+    nqp::nativecallrefresh(self);
```
Comment (FROGGS) [inline lib/XML/LibXML/Attr.pm]: does the call to refresh actually make a difference?

File: lib/XML/LibXML/Attr.pm [hunk 3]
```diff
@@ -25,14 +31,46 @@ method value() {
...
+#     several classes.
+
+method isSameNode($n) {
+    # cw: Maybe.
+
+    # cw: Not. The issue here is that P6 is creating these objects from 
+    #     the backing C-pointer which is lost in the shuffle. So two objects
+    #     can be created from the -same- pointer but have no sense of 
+    #     equivalence (unless... possibly... the repr is 'CPointer')
+    return self =:= $n;
```
Comment (FROGGS) [inline lib/XML/LibXML/Attr.pm]: we can perhaps get our hands at the memory address the pointer points to and compare that...
Would that help?

File: lib/XML/LibXML/Attr.pm [hunk 4]
```diff
@@ -25,14 +31,46 @@ method value() {
...
+}
+
+method ownerDocument {
+    return nativecast(::('XML::LibXML::Document'), self.doc);
```
Comment (FROGGS) [inline lib/XML/LibXML/Attr.pm]: we should try to keep these indirect lookups at a minimum

File: lib/XML/LibXML/CStructs.pm [hunk 5]
```diff
@@ -2,11 +2,15 @@ use v6;
 use nqp;
 use NativeCall;
 
+constant XML_XML_NAMESPACE = "http://www.w3.org/XML/1998/namespace";
+
+my $LIBXML_VERSION := cglobal('libc.so.6', 'LIBXML_VERSION', int32);
```
Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: instead of `'libc.so.6'` you could also pass `Str`

File: lib/XML/LibXML/CStructs.pm [hunk 6]
```diff
@@ -113,6 +119,27 @@ my class xmlError is export(:types) {
     has OpaquePointer $.node; # the node in the tree
 }
 
+my class xmlElement is export(:types) {
```
Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: We need to test that structure in t/01-structure-sizes.t.

File: lib/XML/LibXML/CStructs.pm [hunk 7]
```diff
@@ -132,6 +159,34 @@ my class xmlNode is export(:types) {
     has uint16 $.line;
     #~ unsigned short	extra	: extra data for XPath/XSLT
     has uint16 $.extra;
+
+
+    # cw: Looks like having methods in these classes will be unavoidable unless we 
+    #     decide to make the attributes rw... which might be workable as long as 
+    #     we can protect them in descendent classes
+    method setNext(xmlNodePtr $n) {
```
Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: yes, I'd favour rw-attrs too

File: lib/XML/LibXML/Document.pm [hunk 8]
```diff
@@ -45,7 +46,10 @@ method doctype is aka<type> {
     #~ This read-only property is an object that implements the DOMImplementation interface.
 
 #| This read-only property is an object that implements the Element interface.
-method documentElement is aka<root> {
+method documentElement
```
Comment (FROGGS) [inline lib/XML/LibXML/Document.pm]: What do you think of these aka-methods at all? More and more I think we should stick to the longer more xmllish names...

File: lib/XML/LibXML/Document.pm [hunk 9]
```diff
@@ -290,16 +294,41 @@ multi method new-elem-ns($name, $uri) {
     self.new-elem-ns($name => Str, $uri)
 }
 
-multi method new-attr(Pair $kv) is aka<createAttribute> {
+method createElementNS($_uri, $_name) {
+    if ! testNodeName($_name) {
```
Comment (FROGGS) [inline lib/XML/LibXML/Document.pm]: better written as: unless testNodeName($_name) {

File: lib/XML/LibXML/Document.pm [hunk 10]
```diff
@@ -341,3 +374,37 @@ method new-cdata-block(Str $cdata) {
...
+        die "setDocumentElement: ELEMENT node required";
+        # cw: To properly handle .resume in CATCH {}
+        return;
+    }
+
+    domImportNode(self, $elem, 1, 1);
+    my $oelem = xmlDocGetRootElement(self);
+    if (!$oelem.defined || !$oelem._private.defined) {
+        xmlDocSetRootElement(self, $elem);
+    } else {
```
Comment (FROGGS) [inline lib/XML/LibXML/Document.pm]: No cuddled else please :o)

File: lib/XML/LibXML/Dom.pm [hunk 11]
```diff
@@ -0,0 +1,281 @@
+use v6.c;
+
+use XML::LibXML::CStructs :types;
+
+class XML::LibXML::xmlNs is xmlNs {
+
+	method setNext(xmlNs $n) {
```
Comment (FROGGS) [inline lib/XML/LibXML/Dom.pm]: Are there hard tabs or is that just because of the github diff view? Please stick to four space chars either way.

File: lib/XML/LibXML/Dom.pm [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+	sub _domReconcileNsAttr(xmlAttrPtr $a, xmlNsPtr $unused) {
+		return unless $a.defined;
+
+		my $attr = nativecast(xmlAttr, $a);
+
+		my xmlNodePtr $tree = $attr.parent;
+		my xmlNode $tree_o = nativecast(xmlNode, $tree);
+
+		return if ! $tree.defined;
```
Comment (FROGGS) [inline lib/XML/LibXML/Dom.pm]: no whitespace between postfix and term please

File: lib/XML/LibXML/Attr.pm (hunk 2 above)
Comment (Xliff) [inline lib/XML/LibXML/Attr.pm]: Actually, it did.

File: lib/XML/LibXML/Attr.pm (hunk 3 above)
Comment (Xliff) [inline lib/XML/LibXML/Attr.pm]: Yes. If that doesn't solve it, I don't know what will.

File: lib/XML/LibXML/Attr.pm (hunk 4 above)
Comment (Xliff) [inline lib/XML/LibXML/Attr.pm]: Agreed. I only use them to prevent circular "use" references.

Maybe I can break some of these routines out to other areas where they are not necessary.

File: lib/XML/LibXML/Dom.pm [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+			if ($attr.ns.name.defined && $attr.ns.name eq "xml") {
+				$ns = xmlSearchNsByHref($tree.doc, $tree, XML_XML_NAMESPACE);
+				$attr.setNs($ns);
+				return
+			} else {
+				$ns = xmlSearchNs($tree_o.doc, $tree_o.parent, $attr.ns.name);
+			}
+
+			my $nso = nativecast(xmlNs, $ns);
+			if [&&](
```
Comment (FROGGS) [inline lib/XML/LibXML/Dom.pm]: why the fancy reduce op here? and old-school `if A && B && C` will do, and the &&'s look nicer. (you spot the begging of the block more easily)

File: lib/XML/LibXML/CStructs.pm (hunk 5 above)
Comment (Xliff) [inline lib/XML/LibXML/CStructs.pm]: Actually, that _doesn't work_ because libxml2 does not have a function that expressly returns the version. They have routines for the parser version (maybe that would work) and the SAX version, but not the library version.

LIBXML_VERSION is actually a macro... which doesn't help us in Nativecall.

File: lib/XML/LibXML/Dom.pm [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+				}
+			}
+		}
+	}
+
+	#sub _domReconcileNs(xmlNode $tree, xmlNs $unused) {
+	sub _domReconcileNs($tree, $unused) {
+		sub xmlCopyNamespace(xmlNs) returns xmlNs is native('xml2') { * }
+
+	     if  $tree.ns !~~ Nil     &&
```
Comment (FROGGS) [inline lib/XML/LibXML/Dom.pm]: I prefer:
```if FOO
&& BAR {
```

File: lib/XML/LibXML/CStructs.pm (hunk 6 above)
Comment (Xliff) [inline lib/XML/LibXML/CStructs.pm]: xmlElement? Sure thing.

File: lib/XML/LibXML/Dom.pm [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+	         ($tree.type == XML_ELEMENT_NODE || $tree.type == XML_ATTRIBUTE_NODE)
+	     {
+	         my $ns = xmlSearchNs($tree.doc, $tree.parent, $tree.ns.uri);
+	         if [&&](
+         		$ns.defined,
+	            $ns.href.defined,
+	            $tree.ns.href.defined,
+	            $ns.href eq $tree.ns.href
+	         ) {
+				$unused = _domAddNsChain($unused, $tree.ns)
```
Comment (FROGGS) [inline lib/XML/LibXML/Dom.pm]: consistent indentation please... makes it easier to spot blocks

File: lib/XML/LibXML/CStructs.pm (hunk 7 above)
Comment (Xliff) [inline lib/XML/LibXML/CStructs.pm]: Should this be done for all of the common xmlNodePtr attributes?

File: lib/XML/LibXML/Document.pm (hunk 8 above)
Comment (Xliff) [inline lib/XML/LibXML/Document.pm]: I think we can use them since aliases are also done in p5 XML::LibXML. I think that we should try to go for as much of a drop in replacement where it makes sense. And document well where things deviate. I've been using the XS code from p5's XML::LibXML as a guide and aliasing is heavily used, there.

File: lib/XML/LibXML/Document.pm (hunk 9 above)
Comment (Xliff) [inline lib/XML/LibXML/Document.pm]: Noted! ;)

File: lib/XML/LibXML/Node.pm [hunk 16]
```diff
@@ -108,25 +112,422 @@ role XML::LibXML::Nodish does XML::LibXML::C14N {
...
+        if $attr_ns.defined && $attr_ns.chars && ! $ns.defined {
+            die "bad ns attribute!";
+            return            
+        }
+
+        xmlSetNsProp(self.getNode(), $ns, $localname, $val);
+    }
+
+    method hasAttribute($a) {
+        my $ret = domGetAttrNode(
```
Comment (FROGGS) [inline lib/XML/LibXML/Node.pm]: that short function fits in one line

File: lib/XML/LibXML/Document.pm (hunk 10 above)
Comment (Xliff) [inline lib/XML/LibXML/Document.pm]: LOL! First time I've heard that term. That's an old coding style I adopted which I am happy to discard. Means a lot of changes though. I will knock them out as I find them.

File: lib/XML/LibXML/Node.pm [hunk 17]
```diff
@@ -108,25 +112,422 @@ role XML::LibXML::Nodish does XML::LibXML::C14N {
...
+
+        $attr_ns := Str if !$_ns.defined || !$attr_ns.chars;
+
+        my xmlAttr $attr = nativecast(
+            xmlAttr,
+            xmlHasNsProp(self.getNode(), $name, $attr_ns)
+        );
+
+        return ($attr.defined && $attr.type == XML_ATTRIBUTE_NODE) ??
+            True !! False;
```
Comment (FROGGS) [inline lib/XML/LibXML/Node.pm]: is the Bool of the condition not Bool enough? ;o)

File: lib/XML/LibXML/Dom.pm (hunk 11 above)
Comment (Xliff) [inline lib/XML/LibXML/Dom.pm]: There may be hard tabs. I have recently been replacing them as I find them. 

Another old coding habit because I hated hitting right arrow through code 3 indents deep. Editors are smarter now, so it's not necessary anymore.

Old habits and all that.

File: t/06elements-port.t (hunk 1 above)
Comment (Xliff) [inline t/06elements-port.t]: Noted!

File: lib/XML/LibXML/Node.pm [hunk 18]
```diff
@@ -108,25 +112,422 @@ role XML::LibXML::Nodish does XML::LibXML::C14N {
...
+            $ns = xmlSearchNs(self.doc, self.getNode(), Str);
+            if ($ns.defined && $ns.uri.defined && $ns.uri.chars) {
+                $ret = 0;
+            } elsif $flag {
+                xmlSetNs(self.getNode(), xmlNs);
+                $ret = 1;
+            } else {
+                $ret = 0;
+            }
+        } elsif $flag {
```
Comment (FROGGS) [inline lib/XML/LibXML/Node.pm]: What the heck is `$flag`? we should try to avoid P5's magic numbers... let's einer use enums or named args for that purpose.

File: t/06elements-port.t [hunk 19]
```diff
@@ -0,0 +1,555 @@
...
+<!ATTLIST root name NMTOKEN #IMPLIED>
+<!ENTITY ent "ENT">
+]>
+EOF
+
+    my $ns = 'urn:xx';
+    my $xml_nons = '<root foo="&quot;bar&ent;&quot;" xmlns:a="$ns"/>';
+    my $xml_ns = '<root xmlns="$ns" xmlns:a="$ns" foo="&quot;bar&ent;&quot;"/>';
+
+    # TEST:$xml=2;
```
Comment (Xliff) [inline t/06elements-port.t]: Tobias,

This is where things start to go off of the rails. Particularly when it comes to entity replacement.

File: lib/XML/LibXML/Node.pm [hunk 20]
```diff
@@ -108,25 +112,422 @@ role XML::LibXML::Nodish does XML::LibXML::C14N {
...
+            $ns = xmlSearchNs(self.doc, self.getNode(), $_prefix);
+
+            if ($ns.defined) {
+                if $ns.uri eq $_uri {
+                    $ret = 1;
```
Comment (FROGGS) [inline lib/XML/LibXML/Node.pm]: setting `$ret` here seems to be a noop

File: lib/XML/LibXML/Parser.pm [hunk 21]
```diff
@@ -39,6 +40,11 @@ method parse(Str:D $str, Str :$uri, :$flags = self.html == 1 ?? HTML_PARSE_RECOV
     $doc
 }
 
+method parse_string($str) {
```
Comment (FROGGS) [inline lib/XML/LibXML/Parser.pm]: let's use a dash here

File: lib/XML/LibXML/Subs.pm [hunk 22]
```diff
@@ -37,8 +37,21 @@ sub xmlC14NDocDumpMemory(xmlDoc, xmlNodeSet, int32, CArray[Str], int32, CArray[S
...
 sub xmlKeepBlanksDefault(int32)              returns int32      is native('xml2') is export { * }
 sub xmlIsBlankNode(xmlNode)                  returns int32      is native('xml2') is export { * }
-sub xmlUnlinkNode(xmlNode)                                      is native('xml2') is export { * }
+sub xmlUnlinkNode(xmlNodePtr)                                   is native('xml2') is export { * }
 sub xmlUnsetProp(xmlNode, Str)               returns int32      is native('xml2') is export { * }
 #~ multi xmlChildElementCount(xmlNode)          returns ulong      is native('xml2') is export { * }
 #~ multi xmlChildElementCount(xmlDoc)           returns ulong      is native('xml2') is export { * }
 sub xmlEncodeEntitiesReentrant(xmlDoc, Str)  returns Str        is native('xml2') is export { * }
+
+sub xmlHasNsProp(xmlNode, Str, Str)
```
Comment (FROGGS) [inline lib/XML/LibXML/Subs.pm]: why the newlines?

File: t/06elements-port.t [hunk 23]
```diff
@@ -0,0 +1,555 @@
...
+my $attvalue2 = "b";
+my $attname3  = "C";
+
+# TEST:$badnames=4;
+my @badnames= ("1A", "<><", "&", "-:");
+
+# 1. bound node
+my $doc = XML::LibXML::Document.new();
+my $elem = $doc.createElement( $foo );
+# TEST
```
Comment (FROGGS) [inline t/06elements-port.t]: are these `#TEST` or any value?

File: t/06elements-port.t [hunk 24]
```diff
@@ -0,0 +1,555 @@
...
+is $elem.getAttribute($attname3), "", 'second attribute is also blank';
+
+# cw: Do we really need enclosing block?
+{
+    for @badnames -> $name {
+        try {
+            my $caught = 0;
+            CATCH {
+                # TEST*$badnames
+                ok True, "setAttribute throws an exception for '$name'";
```
Comment (FROGGS) [inline t/06elements-port.t]: dont we have `pass()` and `flunk()` or something?

File: t/06elements-port.t [hunk 25]
```diff
@@ -0,0 +1,555 @@
...
+# TEST
+is 
+    $tattr.name, "{$prefix}:{$foo}", 
+    'name includes proper namespace';
+# TEST
+is $tattr.value, $attvalue2, 'attribute node has proper value';
+
+$elem.removeAttributeNode( $tattr );
+# TEST
+ok
```
Comment (FROGGS) [inline t/06elements-port.t]: you can use `nok` here

File: t/06elements.t [hunk 26]
```diff
@@ -0,0 +1,605 @@
...
+    ok $elem, 'created element';
+    # TEST
+    is $elem.tagName, $foo, 'element has correct tag name';
+
+    {
+        for @badnames -> $name {
+            try { 
+                my $caught = 0;
+                CATCH {
+                    ok $@, "setNodeName throws an exception for $name";
```
Comment (FROGGS) [inline t/06elements.t]: this will result in a dynamic number of total tests in this file.

File: lib/XML/LibXML/Attr.pm (hunk 2 above)
Comment (FROGGS) [inline lib/XML/LibXML/Attr.pm]: :+1:

File: lib/XML/LibXML/Attr.pm (hunk 3 above)
Comment (FROGGS) [inline lib/XML/LibXML/Attr.pm]: we should be able to cast both sides to Pointer, and then: `return +$p1 == +$p2`

File: lib/XML/LibXML/Attr.pm (hunk 4 above)
Comment (FROGGS) [inline lib/XML/LibXML/Attr.pm]: Yes. I didn't mean it as a "please try to change it now", it is just something we have to keep in mind to not make it slower than it has to be.

File: lib/XML/LibXML/CStructs.pm (hunk 5 above)
Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: Ahh, so that does not work either way... sad.

File: lib/XML/LibXML/CStructs.pm (hunk 6 above)
Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: Every structure we use... if the P6 class definition is incorrect, we wont allocate enough bytes and the C lib might access unallocated memory and will explode. if our class definition is bigger byte-wise we will access mem taht the C lib might not have allocated.

File: lib/XML/LibXML/CStructs.pm (hunk 7 above)
Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: either for all or for none, no? I'd like to see an interface where you dont have to remember what attrs are writeable and what attrs have setters.

File: lib/XML/LibXML/Document.pm (hunk 10 above)
Comment (FROGGS) [inline lib/XML/LibXML/Document.pm]: no problem, would not be a stopper for me to merge your code with this left in :o)
(you'd just have to expect that I format the code how I prefer it... I'm probably picky in that regard.)

Comment (Xliff) [inline lib/XML/LibXML/Document.pm]: LOL! I'd do the same. I've learned the hard way that its best to conform to the coding style of the lead dev.

File: lib/XML/LibXML/Dom.pm (hunk 13 above)
Comment (Xliff) [inline lib/XML/LibXML/Dom.pm]: Was originally done that way. It's a matter of preference, I guess. Someone on #perl6 mentioned this was better when using many expressions, so I re-wrote it this way. Switching it back is no problem.

File: lib/XML/LibXML/Dom.pm (hunk 14 above)
Comment (Xliff) [inline lib/XML/LibXML/Dom.pm]: kk - I find that harder to read, but that's just me. =)

File: lib/XML/LibXML/Dom.pm (hunk 15 above)
Comment (Xliff) [inline lib/XML/LibXML/Dom.pm]: Pfft. I didn't catch that.

File: lib/XML/LibXML/Node.pm [hunk 27]
```diff
@@ -108,25 +112,422 @@ role XML::LibXML::Nodish does XML::LibXML::C14N {
             }
         }
     }
+
+    method isSameNode($n) {
+        # cw: Maybe.
```
Comment (Xliff) [inline lib/XML/LibXML/Node.pm]: "Maybe this will work" -- it didn't, so....

File: lib/XML/LibXML/Node.pm (hunk 16 above)
Comment (Xliff) [inline lib/XML/LibXML/Node.pm]: Yes. In case you haven't noticed, I prefer to read vertically rather than scrolling horizontally. Makes things harder to read. Breaking parameters (especially for long functions) into their own lines also helps. That's one style I will keep in my own code. 

This code is short enough, yes. Originally it was longer.

File: lib/XML/LibXML/Node.pm (hunk 17 above)
Comment (Xliff) [inline lib/XML/LibXML/Node.pm]: LOL - I must have been lacking caffeine when I wrote that.

File: lib/XML/LibXML/Node.pm (hunk 16 above)
Comment (FROGGS) [inline lib/XML/LibXML/Node.pm]: Yes, I totally agree that we should split up calls/expressions that are actually too long.

File: lib/XML/LibXML/Node.pm (hunk 18 above)
Comment (Xliff) [inline lib/XML/LibXML/Node.pm]: This entire function got a WTF from me. I basically ported it straight from the XS code and then tried to simplify it. This is probably one of the situations where we diverge from what is in XML::LibXML.

I will look into this one again and see if it can be rewritten concisely.

File: lib/XML/LibXML/Node.pm (hunk 20 above)
Comment (Xliff) [inline lib/XML/LibXML/Node.pm]: See above comment.

File: lib/XML/LibXML/Parser.pm (hunk 21 above)
Comment (Xliff) [inline lib/XML/LibXML/Parser.pm]: =)

File: t/06elements-port.t (hunk 23 above)
Comment (Xliff) [inline t/06elements-port.t]: Vestigial comments from the original code. They serve no purpose, that I can tell.
I will gladly excise them.

File: t/06elements-port.t (hunk 24 above)
Comment (Xliff) [inline t/06elements-port.t]: Yeah. Would probably be better here.

File: lib/XML/LibXML/Node.pm (hunk 18 above)
Comment (FROGGS) [inline lib/XML/LibXML/Node.pm]: Btw, I would not be surprised if we find lots of bugs in P5's XML::LibXML. I already spotted some and the original author is not interested in fixing them... But we will here :o)

File: t/06elements-port.t (hunk 25 above)
Comment (Xliff) [inline t/06elements-port.t]: Fair 'nuff.

File: t/06elements.t (hunk 26 above)
Comment (Xliff) [inline t/06elements.t]: Dynamic, but not indeterminate. I think this is accounted for in the original file. There's nothing here that will cause the code to not complete the loop. The number of tests will be the same each time you run this suite. We can unroll the loop, but that just means more code.

Comment (Xliff) [inline t/06elements.t]: Another thing. I should probably remove this file and replace it with t/06elements-port.t -- that was the intent. It is completely unnecessary since we still have the original one in t/to-be-ported/

Comment (FROGGS) [inline t/06elements.t]: I am thinking that if the `CATCH` was executed that we end up with two tests per iteration, otherwise one.
So I think that either the resume needs to go, or the test in the catch needs to go or the second test needs to happen conditionally.
I think I believe the second way might be the best.

Comment (FROGGS) [inline t/06elements.t]: ohh, we will remove the original files anyway...

File: lib/XML/LibXML/Node.pm (hunk 18 above)
Comment (Xliff) [inline lib/XML/LibXML/Node.pm]: Damn straight! 

(⌐■_■)

File: t/06elements.t (hunk 26 above)
Comment (Xliff) [inline t/06elements.t]: Actually, that is why I've written the block in this way (and why in the .pm files you will find a return after a die(). If there is an exception, $caught will never be set and the nok (soon to be fail) will be the only test. If it succeeds, the ok (soon to be pass) is done and the nok/fail will not execute due to the unless clause. You don't see them in this file because I fixed those in t/06elements-port.t see [(https://github.com/FROGGS/p6-XML-LibXML/pull/1/files/223ef068513e236e2b6df64d91fe75243da87973#diff-c8808174d37f2722a60f4c71e878aacaR68)]

File: lib/XML/LibXML/CStructs.pm (hunk 6 above)
Comment (Xliff) [inline lib/XML/LibXML/CStructs.pm]: Oh, I'm well aware of this after my initial experiences with Nativecall! =)

Looks like later versions of libxml2 will want $.contModel defined if LIBXML_REGEXP_ENABLED is set or not. This is actually a good thing!

File: lib/XML/LibXML/CStructs.pm (hunk 7 above)
Comment (Xliff) [inline lib/XML/LibXML/CStructs.pm]: Common rw attributes it is! =)

**Comment by Xliff** [comment]
Another question: Should the XML::LibXML::Nodish and the XML::LibXML::Common roles be combined?

**Comment by Xliff** [comment]
I'd be happier if there were a way to minimize the number of calls to nativecast() that this commit forced? Is there a better way to handle this?

File: lib/XML/LibXML/CStructs.pm (hunk 6 above)
Comment (Xliff) [inline lib/XML/LibXML/CStructs.pm]: OK. xmlElement is now passing, but xmlSAXHandler and xmlXPathContext didn't. Do you want a separate PR for these?

Comment (FROGGS) [inline lib/XML/LibXML/CStructs.pm]: Let's do these in a separate step, yes.

**Comment by FROGGS** [comment]
We can only get rid of casts if the attributes were of the right type... though, that'd mean we could not have low level C-ish classes and high level P6 classes, but both mixed up.
I'm not sure if I want to do that step...

**Comment by FROGGS** [comment]
Mergin XML::LibXML::Nodish and XML::LibXML::Common sounds sane I guess.

**Comment by Xliff** [comment]
Excellent! Will push that, shortly.

**Comment by FROGGS** [comment]
Merged as: https://github.com/FROGGS/p6-XML-LibXML/commit/0dc9a78a1b99f9d55fb8e9e54a86129adac26a28

**Comment by FROGGS** [comment]
Thank you!

**Comment by Xliff** [comment]
Anytime! Now to figure out why the later tests are failing! =)
